### PR TITLE
[EditableText] preserve selection/composition range on unfocus 

### LINKED ISSUE #86752: Preserve text field's caret position across focus changes
Currently we set the selection and the composing range of a text field to empty when it loses focus (https://github.com/flutter/flutter/issues/10911):
https://github.com/flutter/flutter/blob/b0e51db8d5f0d29603373e34929b8c501bba5354/packages/flutter/lib/src/widgets/editable_text.dart#L2452

## Steps to Reproduce

But on Android if you: 
1. go to a login screen
2. type something in the password field, and move the caret to the middle of the string, 
3. select the username field (the caret should disappear), and use the "next" button on the virtual keyboard to focus the password field again
4. As the password regains focus the caret re-appears, and it should be where it was. 

GTK seems to have the same behavior, except it doesn't collapse the selection while android does (maybe it's the IME?). When you press tab to refocus an input field that previously has a non-collapsed selection, the selection will re-appear. 

Dropping the selection also notifies the text field's `TextEditingController`s listeners (https://github.com/flutter/flutter/issues/79982). Not commonly expected.

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
Related Issue: #79982
Fixes https://github.com/flutter/flutter/issues/86752

Let the selection/composition range of an `EditableText` survive through
unfocus, and allows the original selection/composition range to be
restored when the field regains focus.

### Demo
Unfortunately the selection handle (and the context menu) is gone when a text field regains focus. This probably needs a new issue since the current implementation of `_handleSelectionChanged` (and the corresponding logic in the gesture detector callbacks) is really convoluted.

https://user-images.githubusercontent.com/31859944/126536564-e7155776-a481-4042-afbd-dc2415712d82.mp4

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by LongCatIsLooong** [comment]
@chunhtai I think you mentioned the desktop accessibility implementation specifically use `(-1, -1)` for something. Is that going to be a problem if we don't move the selection and the composing range upon unfocus?

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 1]
```diff
@@ -2687,7 +2688,9 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
                 minLines: widget.minLines,
                 expands: widget.expands,
                 strutStyle: widget.strutStyle,
-                selectionColor: widget.selectionColor,
+                // Special casing for SelectableText, show the selection even
+                // when the field isn't focused
+                selectionColor: (_hasFocus || widget.readOnly) ? widget.selectionColor : null,
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/editable_text.dart:2693]: The `buildTextSpan` method has something similar (the composing region is not painted if the widget is `readOnly`). But this is a little bit too magical for a low level widget.

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 2]
```diff
@@ -3875,6 +3875,8 @@ void main() {
     ));
 
     assert(focusNode.hasFocus);
+    // Autofocus has an one frame delay.
+    await tester.pump();
```
Comment (LongCatIsLooong) [inline packages/flutter/test/widgets/editable_text_test.dart:3879]: Before the change the underline was shown even when the text field is not focused. That's why this test was passing without the additional `pump`.

**Comment by chunhtai** [comment]
> @chunhtai I think you mentioned the desktop accessibility implementation specifically use `(-1, -1)` for something. Is that going to be a problem if we don't move the selection and the composing range upon unfocus?

That should be fine.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 3]
```diff
@@ -2336,7 +2336,7 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
 
   void _onCursorColorTick() {
     renderEditable.cursorColor = widget.cursorColor.withOpacity(_cursorBlinkOpacityController.value);
-    _cursorVisibilityNotifier.value = widget.showCursor && _cursorBlinkOpacityController.value > 0;
+    _cursorVisibilityNotifier.value = widget.showCursor && _hasFocus && _cursorBlinkOpacityController.value > 0;
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/editable_text.dart:2339]: I guess the caret was blinking when the field had a selection even when it wasn't focused.

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 1 above)
Comment (justinmc) [reply packages/flutter/lib/src/widgets/editable_text.dart:2693]: Is it ever possible for a SelectableText to show its selection while another EditableText has focus?

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 4]
```diff
@@ -3848,6 +3848,51 @@ void main() {
...
+    // Everything's just formated the same way.
+    expect((renderEditable.text! as TextSpan).text, 'text composing text');
+    expect(renderEditable.text!.style!.decoration, isNull);
+
+    focusNode.requestFocus();
+    await tester.idle();
+    await tester.pump();
+    expect((renderEditable.text! as TextSpan).children, isNotNull);
+
+
```
Comment (justinmc) [inline packages/flutter/test/widgets/editable_text_test.dart:3887]: Nit: Extra newline here.

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 5]
```diff
@@ -3875,6 +3920,8 @@ void main() {
     ));
 
     assert(focusNode.hasFocus);
+    // Autofocus has an one frame delay.
```
Comment (justinmc) [inline packages/flutter/test/widgets/editable_text_test.dart:3923]: Nit: "an" => "a"

**Review by justinmc — APPROVED** [review]
LGTM 👍  Just one question and some nits.

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 3 above)
Comment (1972chris) [reply packages/flutter/lib/src/widgets/editable_text.dart:2345]: Any suggestions on process of recovering data with this issue

**Review by LongCatIsLooong — COMMENTED** [review]
@justinmc I moved the focus selection highlight logic to high-level text fields. Could you take a look again?

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 6]
```diff
@@ -884,6 +884,11 @@ class _CupertinoTextFieldState extends State<CupertinoTextField> with Restoratio
       _controller!.dispose();
       _controller = null;
     }
+
+    if (widget.focusNode != oldWidget.focusNode) {
+      (oldWidget.focusNode ?? _focusNode)?.removeListener(_handleFocusChanged);
+      (widget.focusNode ?? _focusNode)?.addListener(_handleFocusChanged);
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_field.dart:890]: I guess we weren't swapping these properly before 😁

File: packages/flutter/lib/src/material/text_field.dart [hunk 7]
```diff
@@ -1239,7 +1255,8 @@ class _TextFieldState extends State<TextField> with RestorationMixin implements
           maxLines: widget.maxLines,
           minLines: widget.minLines,
           expands: widget.expands,
-          selectionColor: selectionColor,
+          // Only show the selection highlight when the text field is focused.
```
Comment (justinmc) [inline packages/flutter/lib/src/material/text_field.dart:1258]: Super nit: There was no corresponding comment in CupertinoTextField here.

**Review by justinmc — APPROVED** [review]
Changes LGTM minus a potential scuba failure, happy to look again when that's fixed.

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 6 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/cupertino/text_field.dart:890]: The `_handleFocusChange` listener is added in this PR. We didn't have to remove/add listeners before.

**Pull request merged**